File: src/treemapper/diffctx/mode.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum


class ScoringMode(Enum):
    HYBRID = "hybrid"
    PPR = "ppr"
    EGO = "ego"


@dataclass(frozen=True)
class PipelineConfig:
    discovery: str
    scoring: str
    low_relevance: bool
    bm25_top_k: int
    ego_depth: int
    ppr_alpha: float

    @classmethod
    def from_mode(cls, mode: ScoringMode, n_candidate_files: int = 0) -> PipelineConfig:
        if mode == ScoringMode.PPR:
            return cls(
                discovery="ensemble",
                scoring="ppr",
                low_relevance=False,
                bm25_top_k=1,
                ego_depth=1,
                ppr_alpha=0.60,
            )
        if mode == ScoringMode.EGO:
            return cls(
                discovery="ensemble",
                scoring="ego",
                low_relevance=False,
                bm25_top_k=1,
                ego_depth=2,
                ppr_alpha=0.60,
            )
        is_large = n_candidate_files > 50
        return PipelineConfig(
            discovery="ensemble" if is_large else "default",
            scoring="ego" if is_large else "ppr",
            low_relevance=not is_large,
            bm25_top_k=1 if is_large else 0,
            ego_depth=2 if is_large else 1,
            ppr_alpha=0.60,
        )

File: diffctx/src/mode.rs
use crate::config::limits::PPR;
use crate::config::mode::mode as mode_config;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ScoringMode {
    Ppr,
    Ego,
    Bm25,
}

impl ScoringMode {
    pub fn from_str(s: &str) -> Result<Self, String> {
        match s.to_lowercase().as_str() {
            "ppr" => Ok(Self::Ppr),
            "ego" => Ok(Self::Ego),
            "bm25" => Ok(Self::Bm25),
            other => Err(format!(
                "unknown scoring_mode '{other}': expected one of ppr|ego|bm25"
            )),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DiscoveryKind {
    Default,
    Ensemble,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ScoringKind {
    Ppr,
    Ego,
    Bm25,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ObjectiveMode {
    Submodular,
    BoltzmannModular,
}

impl ObjectiveMode {
    pub fn from_str(s: &str) -> Self {
        match s.to_lowercase().as_str() {
            "boltzmann" | "boltzmann_modular" | "modular_boltzmann" => Self::BoltzmannModular,
            _ => Self::Submodular,
        }
    }
}

#[derive(Debug, Clone)]
pub struct PipelineConfig {
    pub discovery: DiscoveryKind,
    pub scoring: ScoringKind,
    pub objective: ObjectiveMode,
    pub low_relevance_filter: bool,
    pub bm25_top_k: usize,
    pub ego_depth: usize,
    pub ppr_alpha: f64,
}

impl PipelineConfig {
    pub fn from_mode(mode: ScoringMode) -> Self {
        let m = mode_config();
        match mode {
            ScoringMode::Ppr => Self {
                discovery: DiscoveryKind::Ensemble,
                scoring: ScoringKind::Ppr,
                low_relevance_filter: false,
                bm25_top_k: m.bm25_top_k_primary,
                ego_depth: m.ego_depth_default,
                ppr_alpha: PPR.alpha,
                objective: ObjectiveMode::Submodular,
            },
            ScoringMode::Ego => Self {
                discovery: DiscoveryKind::Ensemble,
                scoring: ScoringKind::Ego,
                low_relevance_filter: false,
                bm25_top_k: m.bm25_top_k_primary,
                ego_depth: m.ego_depth_extended,
                ppr_alpha: PPR.alpha,
                objective: ObjectiveMode::Submodular,
            },
            ScoringMode::Bm25 => Self {
                discovery: DiscoveryKind::Ensemble,
                scoring: ScoringKind::Bm25,
                low_relevance_filter: false,
                bm25_top_k: m.bm25_top_k_off,
                ego_depth: m.ego_depth_default,
                ppr_alpha: PPR.alpha,
                objective: ObjectiveMode::Submodular,
            },
        }
    }
}
